// react-native-builder-bob/lib/__fixtures__/project/code/$exports-input.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
var _exportNames = {
  foo: true,
  bar: true,
  a: true,
  b: true,
  c: true,
  d: true,
  e: true,
  f: true,
  pac: true,
  pak: true,
  pax: true,
  a1: true,
  NativeLibA: true,
  NativeLibB: true,
  NativeLibC: true,
  NativeLibD: true,
  NativeLibE: true,
  NativeLibF: true,
  NativeLibG: true,
  NativeLibH: true,
  NativeLibI: true,
  NativeLibJ: true,
  NativeViewA: true,
  NativeViewB: true,
  NativeViewC: true,
  NativeViewD: true,
  NativeViewE: true,
  NativeViewF: true,
  NativeViewG: true,
  NativeViewH: true,
  NativeViewI: true,
  NativeViewJ: true
};
exports.a = exports.NativeViewJ = exports.NativeViewI = exports.NativeViewH = exports.NativeViewG = exports.NativeViewF = exports.NativeViewE = exports.NativeViewD = exports.NativeViewC = exports.NativeViewB = exports.NativeViewA = exports.NativeLibJ = exports.NativeLibI = exports.NativeLibH = exports.NativeLibG = exports.NativeLibF = exports.NativeLibE = exports.NativeLibD = exports.NativeLibC = exports.NativeLibB = exports.NativeLibA = void 0;
Object.defineProperty(exports, "a1", {
  enumerable: true,
  get: function () {
    return _a3.a;
  }
});
exports.pax = exports.pak = exports.pac = exports.foo = exports.f = exports.e = exports.d = exports.c = exports.bar = exports.b = void 0;
var _a = _interopRequireWildcard(require("a"));
exports.a = _a;
var _b = _interopRequireWildcard(require("./b"));
exports.b = _b;
Object.keys(_b).forEach(function (key) {
  if (key === "default" || key === "__esModule") return;
  if (Object.prototype.hasOwnProperty.call(_exportNames, key)) return;
  if (key in exports && exports[key] === _b[key]) return;
  Object.defineProperty(exports, key, {
    enumerable: true,
    get: function () {
      return _b[key];
    }
  });
});
var _c = _interopRequireWildcard(require("./c"));
exports.c = _c;
var _d = _interopRequireWildcard(require("./d"));
exports.d = _d;
var _e = _interopRequireWildcard(require("./e.story"));
exports.e = _e;
var _f = _interopRequireWildcard(require("../f"));
exports.f = _f;
var _pac = _interopRequireWildcard(require(".."));
exports.pac = _pac;
var _pak = _interopRequireWildcard(require("../"));
exports.pak = _pak;
var _pax = _interopRequireWildcard(require("../index"));
exports.pax = _pax;
var _a3 = require("./a");
var _NativeLibA = _interopRequireWildcard(require("./NativeMyLib"));
exports.NativeLibA = _NativeLibA;
var _NativeLibB = _interopRequireWildcard(require("../NativeMyLib"));
exports.NativeLibB = _NativeLibB;
var _NativeLibC = _interopRequireWildcard(require("./NativeMyLib.ts"));
exports.NativeLibC = _NativeLibC;
var _NativeLibD = _interopRequireWildcard(require("../NativeMyLib.ts"));
exports.NativeLibD = _NativeLibD;
var _NativeLibE = _interopRequireWildcard(require("./NativeMyLib.tsx"));
exports.NativeLibE = _NativeLibE;
var _NativeLibF = _interopRequireWildcard(require("../NativeMyLib.tsx"));
exports.NativeLibF = _NativeLibF;
var _NativeLibG = _interopRequireWildcard(require("./NativeMyLib.js"));
exports.NativeLibG = _NativeLibG;
var _NativeLibH = _interopRequireWildcard(require("../NativeMyLib.js"));
exports.NativeLibH = _NativeLibH;
var _NativeLibI = _interopRequireWildcard(require("./NativeMyLib.jsx"));
exports.NativeLibI = _NativeLibI;
var _NativeLibJ = _interopRequireWildcard(require("../NativeMyLib.jsx"));
exports.NativeLibJ = _NativeLibJ;
var _NativeViewA = _interopRequireWildcard(require("./MyNativeComponent"));
exports.NativeViewA = _NativeViewA;
var _NativeViewB = _interopRequireWildcard(require("../MyNativeComponent"));
exports.NativeViewB = _NativeViewB;
var _NativeViewC = _interopRequireWildcard(require("./MyNativeComponent.ts"));
exports.NativeViewC = _NativeViewC;
var _NativeViewD = _interopRequireWildcard(require("../MyNativeComponent.ts"));
exports.NativeViewD = _NativeViewD;
var _NativeViewE = _interopRequireWildcard(require("./MyNativeComponent.tsx"));
exports.NativeViewE = _NativeViewE;
var _NativeViewF = _interopRequireWildcard(require("../MyNativeComponent.tsx"));
exports.NativeViewF = _NativeViewF;
var _NativeViewG = _interopRequireWildcard(require("./MyNativeComponent.js"));
exports.NativeViewG = _NativeViewG;
var _NativeViewH = _interopRequireWildcard(require("../MyNativeComponent.js"));
exports.NativeViewH = _NativeViewH;
var _NativeViewI = _interopRequireWildcard(require("./MyNativeComponent.jsx"));
exports.NativeViewI = _NativeViewI;
var _NativeViewJ = _interopRequireWildcard(require("../MyNativeComponent.jsx"));
exports.NativeViewJ = _NativeViewJ;
function _getRequireWildcardCache(e) { if ("function" != typeof WeakMap) return null; var r = new WeakMap(), t = new WeakMap(); return (_getRequireWildcardCache = function (e) { return e ? t : r; })(e); }
function _interopRequireWildcard(e, r) { if (!r && e && e.__esModule) return e; if (null === e || "object" != typeof e && "function" != typeof e) return { default: e }; var t = _getRequireWildcardCache(r); if (t && t.has(e)) return t.get(e); var n = { __proto__: null }, a = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var u in e) if ("default" !== u && {}.hasOwnProperty.call(e, u)) { var i = a ? Object.getOwnPropertyDescriptor(e, u) : null; i && (i.get || i.set) ? Object.defineProperty(n, u, i) : n[u] = e[u]; } return n.default = e, t && t.set(e, n), n; }
const foo = exports.foo = "foo";
const bar = exports.bar = "bar";
