// react-native-builder-bob/lib/__tests__/babel.test.js
"use strict";

var _vitest = require("vitest");
var _core = require("@babel/core");
var _nodeFs = _interopRequireDefault(require("node:fs"));
var _nodePath = _interopRequireDefault(require("node:path"));
var _babel = _interopRequireDefault(require("../babel"));
function _interopRequireDefault(e) { return e && e.__esModule ? e : { default: e }; }
_vitest.test.each(['imports', 'exports'])(`adds extension to %s`, async name => {
  const filepath = _nodePath.default.resolve(__dirname, `../__fixtures__/project/code/$${name}-input.ts`);
  const result = await (0, _core.transformFileAsync)(filepath, {
    caller: {
      name: 'test',
      supportsStaticESM: false,
      rewriteImportExtensions: true,
      jsxRuntime: 'automatic',
      codegenEnabled: true
    },
    configFile: false,
    babelrc: false,
    plugins: ['@babel/plugin-syntax-typescript', [_babel.default, {
      extension: 'mjs'
    }]]
  });
  const expected = await _nodeFs.default.promises.readFile(_nodePath.default.resolve(__dirname, `../__fixtures__/project/code/$${name}-output.ts`), 'utf8');
  (0, _vitest.expect)(result?.code).toEqual(expected.trim());
});
