// conventional-recommended-bump/dist/bumper.js
import { ConventionalGitClient, packagePrefix } from '@conventional-changelog/git-client';
import { loadPreset } from 'conventional-changelog-preset-loader';
import { isIterable } from './utils.js';
export { packagePrefix };
const VERSIONS = [
    'major',
    'minor',
    'patch'
];
/**
 * Bump suggester for conventional commits
 */
export class Bumper {
    gitClient;
    preset;
    whatBump;
    tagGetter;
    commitsGetter;
    constructor(cwdOrGitClient = process.cwd()) {
        this.gitClient = typeof cwdOrGitClient === 'string'
            ? new ConventionalGitClient(cwdOrGitClient)
            : cwdOrGitClient;
        this.preset = null;
        this.whatBump = null;
        this.tagGetter = () => this.getLastSemverTag();
        this.commitsGetter = () => this.getCommits();
    }
    getLastSemverTag(params) {
        return this.gitClient.getLastSemverTag(params);
    }
    async *getCommits(params, parserOptions) {
        yield* this.gitClient.getCommits({
            format: '%B%n-hash-%n%H',
            from: await this.tagGetter() || '',
            filterReverts: true,
            ...params
        }, parserOptions);
    }
    async getPreset() {
        const result = await this.preset;
        if (!result) {
            throw Error('Preset is not loaded or have incorrect exports');
        }
        return result;
    }
    /**
     * Load configs from a preset
     * @param preset
     * @returns this
     */
    loadPreset(preset) {
        this.preset = loadPreset(preset);
        this.whatBump = async (commits) => {
            const { whatBump } = await this.getPreset();
            return whatBump(commits);
        };
        this.tagGetter = async () => {
            const { tags } = await this.getPreset();
            return this.getLastSemverTag(tags);
        };
        this.commitsGetter = async function* commitsGetter() {
            const { commits, parser } = await this.getPreset();
            yield* this.getCommits(commits, parser);
        };
        return this;
    }
    /**
     * Set params to get the last semver tag
     * @param paramsOrTag - Params to get the last semver tag or a tag name
     * @returns this
     */
    tag(paramsOrTag) {
        if (typeof paramsOrTag === 'string') {
            this.tagGetter = () => paramsOrTag;
        }
        else {
            this.tagGetter = () => this.getLastSemverTag(paramsOrTag);
        }
        return this;
    }
    commits(paramsOrCommits, parserOptions) {
        if (isIterable(paramsOrCommits)) {
            this.commitsGetter = () => paramsOrCommits;
        }
        else {
            this.commitsGetter = () => this.getCommits(paramsOrCommits, parserOptions);
        }
        return this;
    }
    /**
     * Recommend a bump by `whatBump` function
     * @param whatBump - Function to recommend a bump from commits
     * @returns Bump recommendation
     */
    async bump(whatBump = this.whatBump) {
        if (typeof whatBump !== 'function') {
            throw Error('`whatBump` must be a function');
        }
        const commitsStream = this.commitsGetter();
        const commits = [];
        let commit;
        for await (commit of commitsStream) {
            commits.push(commit);
        }
        let result = await whatBump(commits);
        if (result && typeof result.level === 'number') {
            result.releaseType = VERSIONS[result.level];
        }
        else if (!result) {
            result = {};
        }
        return result;
    }
}
